File: 백준/Silver/16173. 점프왕 쩰리 （Small）/점프왕 쩰리 （Small）.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine().trim());
        int[][] map = new int[N][N];

        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        boolean[][] visited = new boolean[N][N];
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{0, 0});
        visited[0][0] = true;

        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            int r = cur[0], c = cur[1];

            if (r == N - 1 && c == N - 1) {
                System.out.println("HaruHaru");
                return;
            }

            int jump = map[r][c];

            // 오른쪽으로 jump칸
            int nc = c + jump;
            if (nc < N && !visited[r][nc]) {
                visited[r][nc] = true;
                queue.offer(new int[]{r, nc});
            }

            // 아래로 jump칸
            int nr = r + jump;
            if (nr < N && !visited[nr][c]) {
                visited[nr][c] = true;
                queue.offer(new int[]{nr, c});
            }
        }

        System.out.println("Hing");
    }
}

File: 백준/Silver/16173. 점프왕 쩰리 （Small）/점프왕 쩰리 （Small）.kt
fun main() {
    val n = readln().toInt()
    val map = Array(n) { readln().split(" ").map { it.toInt() }.toIntArray() }
    val visited = Array(n) { BooleanArray(n) }
    
    var found = false
    
    fun dfs(x: Int, y: Int) {
        if (x < 0 || x >= n || y < 0 || y >= n) return
        
        if (visited[x][y]) return
        
        visited[x][y] = true
        
        if (map[x][y] == -1) {
            found = true
            return
        }
        
        val jump = map[x][y]
        
        dfs(x, y + jump)
        dfs(x + jump, y)
    }
    
    dfs(0, 0)
    
    println(if (found) "HaruHaru" else "Hing")
}